caddytls/handshake.go

```diff
@@ -159,10 +161,49 @@ func (cfg *Config) getCertDuringHandshake(name string, loadIfNecessary, obtainIf
 	return Certificate{}, fmt.Errorf("no certificate available for %s", name)
 }
 
+// checkIfCertShouldBeObtained checks to see if an on-demand tls certificate
+// should be obtained for a given domain based upon the config settings.  If
+// a non-nil error is returned, do not issue a new certificate for name.
+func (cfg *Config) checkIfCertShouldBeObtained(name string) error {
+	// If the "ask" URL is defined in the config, use to determine if a
+	// cert should obtained
+	if cfg.OnDemandState.CheckUrl != "" {
+		return cfg.checkURLForObtainingNewCerts(name)
+	}
+
+	// Otherwise use the limit defined by the "max_certs" setting
+	return cfg.checkLimitsForObtainingNewCerts(name)
+}
+
+func (cfg *Config) checkURLForObtainingNewCerts(name string) error {
+	client := http.Client{
+		Timeout: 10 * time.Second,
+	}
+
+	checkURL, err := url.Parse(cfg.OnDemandState.CheckUrl)
+	if err != nil {
+		return err
+	}
+
+	query := checkURL.Query()
+	query.Add("domain", name)
+	checkURL.RawQuery = query.Encode()
+
+	resp, err := client.Get(checkURL.String())
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode < 200 || resp.StatusCode > 299 {
+		return fmt.Errorf("Non-200 status code returned (%d)", resp.StatusCode)
```

Comment:
Start the error string with a lower-case letter, since it is usually used in the middle of a sentence when printing to the logs. Also, be a little more descriptive about the implications of this. For example: "certificate for hostname '%s' not allowed, non-2xx status code %d returned"